### ISSUE TITLE
[Feature Request] Add retro_set_raster_poll API Proposal (To Incubate Beam Raced Outputs)

### ISSUE THREAD

**Issue by mdrejhon** [libretro/RetroArch#10758]
We need a very simple per-raster callback function called "retro_set_raster_poll" whose arguments are identical to "retro_set_video_refresh".

Once the centralized hook is done, all emulator modules can just chain it to retro_set_video_refresh by default. This will be default behavior, for beamrace-unaware backwards compatibility

This API is a prerequisite for multiple GitHub items:

- #6984  
- #10757 

For emulator modules that choose to support beamraced outputs,

- Emulator module calls retro_set_raster_poll everytime an emulator scan line is plotted
- The incomplete contents of the emulator framebuffer (complete up to the most recently plotted emulator scanline) is provided. 

This allows supporting beamraced outputs in a simple & cross-platform way.

**Comment by mdrejhon** [comment]
For those people who ask questions about call-blocking behaviours:

Whomever implements this GitHub item don't have to worry about whomever implements beamraced outputs (#6984 or #10757).  This information is only provided to provide call-blocking-behaviour understandings, for emulation sync programming.

**Handler of retro_set_raster_poll is permitted to block briefly (sub refresh)**

Be noted, return from retro_set_raster_poll may be erratic delays (at the microsecond timescales) because the handler (whomever implements #6984 or #10757) may block return for raster interrupt beam racing synchronization purposes.  

The module that handles retro_set_raster_poll may be synchronizing to a hardware-based raster, for example, or a refresh cycle interval (e.g. beamracing a 60Hz output to a 360Hz display in segments), as there are multiple hardware-based and software-based beam raced output approaches.    

So if it is doing 600-frameslice beam racing, and if it about 1/10th refresh ahead of schedule, it might suddenly decide to block for 1/10th of a refresh cycle, before returning.

This is no different from the existing retro_set_video_refresh which can block for 1 full refresh cycle.  

Instead of retro_set_video_refresh blocking for 1 refresh cycle, the combined 200 calls to retro_set_raster_poll for a 200-scanline emulator might potentially have a total sum blockage of one emulator refresh cycle.   Basically it's just retro_set_video_refresh blockages fragmented into many tiny pieces.

These situations may be experienced:

- **retro_set_raster_poll** creates 200 blockages of 1/200th refresh. 
_(Example: beam racing to a NVIDIA VRWorks front buffer)_
- **retro_set_raster_poll** creates 10 blockages of 10/200th refresh, and 190 immediate "returns;"
_(Example use case: VSYNC OFF frameslice beam racing in #6984)_
- **retro_set_raster_poll** creates no blockages (200 immediate returns;)
_(Example use case: beam racing not supported by emulator module)_

The remainder of blocking will occur in retro_set_video_refresh (the normal VSYNC ON blocking behaviour) for whatever **retro_set_raster_poll** didn't block.

For perfect output beamracing, all display output will have been automatically done inside the  **retro_set_raster_poll** with the final afterwards retro_set_video_refresh being NOP (an immediate return;)

**Emulator module doesn't have to worry about how output is beamraced**

The bottom line is the emulator module shouldn't have to care whether the output display was rendered via retro_set_raster_poll or via retro_set_video_refresh

**Comment by mdrejhon** [comment]
There's a command line option (use 3 instead of 1 black frame) to get 240Hz BFI working better.  However, it doesn't work very well with all emulator modules -- the frame timing of RetroArch could use some further improvement, to try to time the frame presentation as micro-second prcise.  

The snap-to behavior of 60Hz refresh cycles (1/60sec ~= 16.7) is much easier than the snap-to behavior of 240Hz refresh cycles (1/240sec ~= 4.2ms) which presents more opportunity for the visible/black frames to accidentally display to the wrong refresh cycles.   

Your computer should also have its power management disabled, because GPU power management (e.g. idling between 60Hz emulator refresh cycles) can be a problem.   

Emulators with command line options to let it execute GPU operations in realtime (e.g. draw rasters to its GPU framebuffer every 1ms) usually work better than emulators that surge-execute at the beginning of its 1/60sec interval.  

This means if one emulator takes only 1-2ms to finish rendering a 60Hz emulated refresh cycles, your real GPU is idling for literally ~15ms until next emulator refresh cycle.  That puts the GPU Power Management to sleep, and produces new timing inaccuracies.  

That's why I recommend #10758 -- "retro_set_raster_poll" API should be added.

By encouraging emulator modules to relay one rendered scanline at a time to RetroArch, this will improve reliability of RetroArch even without frameslice beamracing, because it will keep the GPU's from sleeping, by metering out GPU rendering in tiny time slices (e.g. every 1ms), prevening the GPU from going to sleep between emulator refresh cycles.   So this refactoring by pre-requisite #10758 has some MAJOR spinoff benefits, even without frameslice beam racing.

Even emulators that don't beamrace, but execute their frame rendering in realtime (e.g. run in 1ms increments and render directly to GPU framebuffer without letting GPU power-management itself to sleep) also tends to framepace better on 240Hz monitors without needing VRR help, making other things like monolithic BFI more practical.

Obviously, RunAhead is different (not realtime), but it can still benefit; it just runs a big bunch of frames continually (fast raster scanout to many frames), so it will still be a benefit.  And you can disable processing in retro_set_raster_poll (dummy hooks) for most modules, and just iterate enhancements later into it.   And when 1000Hz OLEDs come, we can forget frameslice beam racing, and simply display 16 incrementally-scanned-out full screen digital refresh cycles per 1 real refresh cycle (much easier to implement low-lag beam racing this way, since full screen refresh cycles are only 1ms).   I'm going to see the 480Hz OLED in the invite-only demo room at CES 2024, so ultra high Hz displays are among us, one can't buy a desktop 27" OLED with a refresh rate less than 240Hz; it merely starts there -- even for productivity.

I'd like to see retro_set_raster_poll (#10758) implemented.  

Special offer: I'll even water-down my existing bounty (#6984) to be paid out only on the mere completion of #10758 if it helps reduce workload.  It's such an important germane improvement that enables a lot of spinoff benefits:

# Benefits Of Just Merely Programming Only #10758

1. Even without beam racing the destination display, it allows slow continuous metering out to GPU for preventing power management stutters
2. Frameslice beam racing (#6984)
3. Ultra high Hz beam racing (480Hz OLEDs allow fully easily displaying 8 partially scanned out emulator frames per 60Hz emulator refresh cycle, and upcoming 1000Hz OLEDs allow fully easily displaying 16 partially scanned out emulator frames per 60Hz emulator refresh cycles)
4. Etc.

cc: @hizzlekizzle

**Comment by hizzlekizzle** [comment]
oh wow, sounds like those new retrotinks are going to be well worth the money

**Comment by mdrejhon** [comment]
I just found out BountySource is insolvement, so ignore that.  I'll put up the money anyway.

# $500 Bounty -- directly from Blur Busters / TestUFO

Inquire within, send to squad [at] blurbusters with subject line "Code Bounty", directly referring to this.  

## Bounty Conditions 

The rule is very simple: Payout will occur when 

1. RetroArch github leads (@hizzlekizzle) closes this issue #10758 
2. At least one emulator module is now calling raster API every raster
3. I authenticate github UserID who made the pull request
4. Expiry date set December 31st, 2024
5. Questions are welcome
